File: 1380. Lucky Numbers in a Matrix/luckyNumbersWithConstantSpace.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
const luckyNumbers  = (matrix) => {
  if (!matrix.length) {
    return null;
  }

  const ROWS = matrix.length;
  const COLS = matrix[0].length;
  let maxOfRowMins = -Infinity;

  for (let r = 0; r < ROWS; r++) {
    const minInRow = Math.min(...matrix[r]);
    maxOfRowMins = Math.max(maxOfRowMins, minInRow);
  }

  for (let c = 0; c < COLS; c++) {
    let currMin = matrix[0][c];
    for (let r = 0; r < ROWS; r++) {
      currMin = Math.max(currMin, matrix[r][c]);
    }
    if (maxOfRowMins === currMin) {
      return [maxOfRowMins];
    }
  }

  return [];
};


const matrix = [
  [3, 7, 8],
  [9, 11, 13],
  [15, 16, 17]
];
console.log('luckyNumbers: ', luckyNumbers(matrix));

File: 1380. Lucky Numbers in a Matrix/luckyNumbersWithConstantSpace.py
from typing import List

class Solution:
  def luckyNumbers (self, matrix: List[List[int]]) -> List[int]:
    if not matrix:
      return None
    
    ROWS, COLS = len(matrix), len(matrix[0])
    maxOfRowMins = float('-inf')

    for r in range(ROWS):
      minInRow = min(matrix[r])
      maxOfRowMins = max(maxOfRowMins, minInRow)

    for c in range(COLS):
      currMax = matrix[0][c]
      for r in range(ROWS):
        currMax = max(currMax, matrix[r][c])
      if currMax == maxOfRowMins:
        return [maxOfRowMins]
    
    return []


matrix = [
  [3, 7, 8],
  [9, 11, 13],
  [15, 16, 17]
]
solution = Solution()
print('luckyNumbers: ', solution.luckyNumbers(matrix))
